Documentation for LD matrix methods (#3416)
tskit-dev/tskit: docs/stats.md

(sec_stats_multi_site)=

## Multi site statistics

(sec_stats_two_locus)=

### Two-locus statistics

The {meth}`~TreeSequence.ld_matrix` method provides an interface to
a collection of two-locus statistics with predefined summary functions (see
{ref}`sec_stats_two_locus_summary_functions`) and `site` and `branch`
{ref}`modes <sec_stats_mode>`. The LD matrix method differs from other
statistics methods in that it provides a unified API with an argument to
specify different two-locus summaries of the data. It otherwise behaves
similarly to most other functions with respect to `sample_sets` and `indexes`.

Two-locus statistics can be computed using two modes, either `site` or
`branch`, and these should be interpreted in the same way as these modes in the
single-site statistics. Within this framework, we also implement polarisation,
but do not expose it to the user, opting to provide statistics that are
polarised where appropriate.

(sec_stats_two_locus_site)=

#### Site

The `site` mode computes two-locus statistics summarized over alleles between
all pairs of specified sites. The default behavior, leaving `sites`

> unspecified, this method will compute a matrix for all pairs of sites, with

Reviewer: ```suggestion
unspecified, will compute a matrix for all pairs of sites, with
```

> rows and columns representing each site in the tree sequence (i.e., an n×n

Reviewer: ```suggestion
one row and column for each site in the tree sequence (i.e., an n x n
```

for both rows and columns or a pair of vectors for the row sites and column
sites separately.

The following computes a matrix of the {math}`r^2` measure of linkage
disequilibrium (LD) computed pairwise between the first 4 sites in the tree
sequence among all samples. In our computations, row sites are used as the row
("left-hand") loci and column sites are used as the column ("right-hand")
locus, and with a single list of sites specified, we obtain a symmetric square
matrix. 

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[0, 1, 2, 3]])
print(ld)
```

The following demonstrates how we can specify the row and column sites
independently of each other. We're specifying 3 columns and 2 rows, which
computes a subset of the matrix shown above.

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[0, 1], [1, 2, 3]])
print(ld)
```

Because we implement two-locus statistics for multi-allelic data, we require
a method for combining the statistical results from each pair of alleles into
one summary for a pair of sites. There two methods for combining results from
multiple alleles, `hap_weighted` and `total_weighted`, which are

> statistic-specific and not chosen by the user:

Reviewer: missing description of hap_weighted and total_weighted here?

Author: I had earlier removed a long description of these to streamline the docs, but agree that I should add a brief explanation of each back in.

(sec_stats_two_locus_branch)=

#### Branch

The `branch` mode computes two-locus statistics between pairs of trees. The
resulting statistic is a summary depending on marginal tree topologies and the
branch lengths of the trees at a pair of positions. To perform this
computation, we consider the counts of all possible two-locus haplotypes that

> may be generated by possible mutations on each pair of branches between the two

Reviewer: ```suggestion
could be generated by mutations on each pair of branches between the two
```

trees, which is summarized using an LD statistic. We then sum each pairwise
comparison weighted by the product of the two branch lengths above the nodes on
each tree.

Similar to the single-site statistics computed in `branch` mode, this results
in a statistic that is proportional to the expected statistic under an infinite

> sites model (with mutation rate 1), conditioned on the pair of trees.

Reviewer: This isn't quite right. At first I was going to just suggest the following:
```
in a statistic that is proportional to the expected statistic under an infinite
sites model. More precisely, suppose we place mutations on each tree as
a Poisson process of rate 1 per unit of branch length, compute
the statistic for each pair of mutations (one on each tree), and add these
up; the `branch` mode is the expected value of the resulting sum.
(Note that unlike the single-site
statistics, these mutations are placed on the trees at a given specific
pair of positions, and so the span of the trees do not affect the result.)
```

However, quick experimenting with `ts.ld_matrix()` to check if this was right pointed out that values can easily be `nan` and so an expected value across mutations is not well-defined. I think we described what this is in the original issue on this?

Reviewer: See more discussion on this below.

The time complexity of this method is quadratic, due to the pairwise

> comparisons of branches from a pair of trees. By default, this method computes

Reviewer: ```suggestion
The time complexity of this method is quadratic in the number of samples,
due to the pairwise comparisons of branches from each pair of trees.
By default, this method computes
```

a symmetric matrix for all pairs of trees, with rows and columns representing
each tree in the tree sequence. Similar to the site method, we can restrict the
output to a subset of trees, either by specifying a vector of positions or
a pair of vectors for row and column positions separately. To select a specific
tree, the specified positions must land in the tree span (`[start, end)`).

In the following, we compute a matrix of expected {math}`r^2` within and
between the first 4 trees in the tree sequence. The tree breakpoints are

> a convenient way to specify those first four trees tree.

Reviewer: ```suggestion
a convenient way to specify those first four trees.
```

```{code-cell} ipython3
breakpoints = ts.breakpoints(as_array=True)
ld = ts.ld_matrix(mode="branch", positions=[breakpoints[[0]], breakpoints[0:4]])
print(ld)
```

(sec_stats_two_locus_sample_sets)=

#### Sample Sets

The two-locus API provides a mechanism by which to subset the samples under
consideration, providing the ability to compute a separate LD matrix for each
sample set or an LD matrix between sample sets. Output dimensions are handled
in the same manner as the rest of the stats API (see
{ref}`sec_stats_output_dimensions`). The two-locus API differs from the rest of
the stats API in that we handle one-way and two-way statistics in the same
function call.

To compute a two-way two-locus statistic, the `index` argument must be
provided. The statistics are selected in the same way (with the `stat`
argument), but we provide a restricted set of two-way statistics (see
{ref}`sec_stats_two_locus_summary_functions_two_way`). The dimension-dropping
rules for the result follow the rest of the tskit stats API in that scalar
indexes will produce a single two-dimensional matrix, while list of indexes
will produce a three-dimensional array, with the outer dimension of length

> equal to the length of the list of indexes.

Reviewer: ```suggestion
rules for the result follow the rest of the tskit stats API in that a single list
or tuple will produce a single two-dimensional matrix, while list of these
will produce a three-dimensional array, with the outer dimension of length
equal to the length of the list.
```

a different meaning from their one-way counterparts. For instance {math}`D_i
D_j` can be thought of as the covariance between two loci when computed for
a haplotype in a single sample set.

Only a subset of our summary functions are two-way statistics (see
{ref}`sec_two_locus_summary_functions_two_way`). Note that the unbiased two-way
statistics expect non-overlapping sample sets, and we do not make any
assertions about the sample sets and assume that `i` and `j` represent disjoint
sets of samples (see also the note in {meth}`~TreeSequence.divergence`).

(sec_stats_two_locus_summary_functions)=

#### Summary Functions

(sec_stats_two_locus_summary_functions_one_way)=

##### One-way

The two-locus summary functions all take haplotype counts and sample set size as
input. Each of our summary functions has the signature
{math}`f(w_{Ab}, w_{aB}, w_{AB}, n)`, converting to haplotype frequencies
{math}`\{p_{Ab}, p_{aB}, p_{AB}\}` where appropriate by dividing by {math}`n`.

`D`
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = p_{ab} - p_{a}p_{b}`

  This statistic is polarised, as the unpolarised result, which averages over

>   allele labelings, is zero. Uses the `total` normalisation method.

Reviewer: below this is called "total weighted" not "normalised"; just checking?

`D_prime`
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = \frac{D}{D_{\max}}`

  Where {math}```

>         D_{\max} = \begin{cases}

Reviewer: ```suggestion
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = \frac{D}{D_{\max}}`,

  where {math}```
        D_{\max} = \begin{cases}
```

            \min\{p_A (1-p_B), p_B (1-p_B)\} & \textrm{if }D>=0 \\
            \min\{p_A p_B, (1-p_B) (1-p_B)\} & \textrm{otherwise}
        \end{cases}```

>   and {math}`D` is defined above.

Reviewer: ```suggestion
  and {math}`D` is defined above. Polarised, `total` weighted.
```

`D2`
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = (p_{ab} - p_{a} p_{b})^2`

>   Unpolarised, total weighted.

Reviewer: ```suggestion
  Unpolarised, `total` weighted.
```

`Dz`
:  {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = D (1 - 2 p_{a})(1 - 2p_{b})`

>   Where {math}`D` is defined above. Unpolarised, total weighted.

Reviewer: ```suggestion
:  {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = D (1 - 2 p_{a})(1 - 2p_{b})`,

  where {math}`D` is defined above. Unpolarised, `total` weighted.
```

Reviewer: (the "where" is part of the same sentence as the equation, not a new sentence, so it's not capitalized)

`pi2`
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = p_{a}p_{b}(1-p_{a})(1-p_{b})`

>   Unpolarised, total weighted.

Reviewer: ```suggestion
  Unpolarised, `total` weighted.
```

`r`
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = \frac{D}{\sqrt{p_{a}p_{b}(1-p_{a})(1-p_{b})}}`

  Where {math}`D` is defined above. Polarised, total weighted.

> 

Reviewer: ```suggestion
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = \frac{D}{\sqrt{p_{a}p_{b}(1-p_{a})(1-p_{b})}}`,

  where {math}`D` is defined above. Polarised, `total` weighted.

```

`r2`
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = \frac{D^{2}}{p_{a}p_{b}(1-p_{a})(1-p_{b})}`

>   Where {math}`D` is defined above. Unpolarised, haplotype weighted.

Reviewer: ```suggestion
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = \frac{D^{2}}{p_{a}p_{b}(1-p_{a})(1-p_{b})}`,

  where {math}`D` is defined above. Unpolarised, `haplotype` weighted.
```

Unbiased two-locus statistics from the Hill-Robertson (1968) system are
computed from haplotype counts. Derivations for these unbiased estimators can
be found in [Ragsdale and Gravel

> (2020)](https://doi.org/10.1093/molbev/msz265). They require at least 4 samples

Reviewer: ```suggestion
computed from haplotype counts. Definitions of these unbiased estimators can
be found in [Ragsdale and Gravel
(2020)](https://doi.org/10.1093/molbev/msz265). They require at least 4 samples
```
Also derivations, but it's more relevant that the defintiions are there?

to be valid and are specified as `stat=D2_unbiased`, `Dz_unbiased`, or

> `pi2_unbiased`.

Reviewer: ```suggestion
to be valid and are specified as `stat="D2_unbiased"`, `"Dz_unbiased"`, or
`"pi2_unbiased"`.
```

(sec_two_locus_summary_functions_two_way)=

(sec_stats_two_locus_summary_functions_two_way)=

##### Two-way

Two-way statistics are indexed by sample sets {math}`i, j` and compute values
using haplotype counts within pairs of sample sets.

`D2`
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = D_i * D_j`

> Where {math}`D` is defined above.

Reviewer: ```suggestion
: {math}`f(w_{Ab}, w_{aB}, w_{AB}, n) = D_i * D_j`,

where {math}`D` is defined above.
```

(sec_stats_multi_site)=

## Multi site statistics

(sec_stats_two_locus)=

### Two-locus statistics

The {meth}`~TreeSequence.ld_matrix` method provides an interface to
a collection of two-locus statistics with predefined summary functions (see
{ref}`sec_stats_two_locus_summary_functions`) and `site` and `branch`
{ref}`modes <sec_stats_mode>`. The LD matrix method differs from other
statistics methods in that it provides a unified API with an argument to
specify different two-locus summaries of the data. It otherwise behaves
similarly to most other functions with respect to `sample_sets` and `indexes`.

Two-locus statistics can be computed using two modes, either `site` or
`branch`, and these should be interpreted in the same way as these modes in the
single-site statistics. Within this framework, statistics may be either
polarised or unpolarised. For statistics that are polarised, we compute
statistic values for pairs of derived alleles. Unpolarised statistics compute
statistics over all possible alleles, derived and ancestral, and the result in
averaged over statistics computed for all pairs of alleles (see weighting

> schemes below). The option for polarisation is not exposed to the user.

Reviewer: ```suggestion
statistic values for pairs of derived alleles. Unpolarised statistics compute
statistics over all pairs of alleles, derived and ancestral. In either case,
the result is averaged over these values, using a weighting
scheme descrbed below. The option for polarisation is not exposed to the user.
```

Reviewer: I'm wondering if this note about polarisation should be down below with the "weighting" schemes. Since isn't polarisation just chosen so that "people get what they expect"? So perhaps bundle those two bits into a "Computational details" section?

matrix where n is the number of sites in the tree sequence). We can also
restrict the output to a subset of sites, either by specifying a single vector
for both rows and columns or a pair of vectors for the row sites and column
sites separately.

The following computes a matrix of the {math}`r^2` measure of linkage
disequilibrium (LD) computed pairwise between the first 4 sites in the tree
sequence among all samples. In our computations, row sites are used as the row
("left-hand") loci and column sites are used as the column ("right-hand")
locus, and with a single list of sites specified, we obtain a symmetric square
matrix. 

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[0, 1, 2, 3]])
print(ld)
```

The following demonstrates how we can specify the row and column sites
independently of each other. We're specifying 3 columns and 2 rows, which
computes a subset of the matrix shown above.

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[0, 1], [1, 2, 3]])
print(ld)
```

Because we allow for two-locus statistics to be computed for multi-allelic
data, we need to be able to combine statistical results from each pair of

> alleles into one summary for a pair of sites. We use two implementations for

Reviewer: ```suggestion
alleles into one summary for a pair of sites. This does not affect biallelic
data (and so this section can be skipped on first reading).
We use two implementations for
```

combining results from multiple alleles: `hap_weighted` and `total_weighted`.
These are statistic-specific and not chosen by the user.

Briefly, consider a pair of sites with {math}`n` alleles at the first locus and

> {math}`m` alleles at the second. Write {math}`f_{i,j}` as the statistic

Reviewer: ```suggestion
{math}`m` alleles at the second. (Whether this includes the ancestral allele
depends on whether the statistic is polarised.) Write {math}`f_{i,j}` as the statistic
```

computed for focal alleles {math}`A_i` and {math}`B_j`, with haplotype weights
{math}`(A_i B_j, A_i b_j, a_i B_j)`, where {math}`a_i` and {math}`b_j` are the
collection of alleles that are not the focal alleles {math}`A_i` or

> {math}`B_j`, respectively. Then the weighting schemes are defined as:

Reviewer: ```suggestion
computed for focal alleles {math}`A_i` and {math}`B_j`.
Then the weighting schemes are defined as:
```
This bit didn't make sense, and I think wasn't used anyhow.

- `hap_weighted`: {math}`\sum_{i=1}^{n}\sum_{j=1}^{m}p(A_{i}B_{j})f_{ij}`,
  where {math}`p(A_{i}B_{j})` is the frequency of haplotype {math}`A_{i}B_{j}`.
  This method was first introduced in [Karlin
  (1981)](https://doi.org/10.1111/j.1469-1809.1981.tb00308.x) and reviewed in
  [Zhao (2007)](https://doi.org/10.1017/S0016672307008634).

- `total_weighted`: {math}`\frac{1}{n m}\sum_{i=1}^{n}\sum_{j=1}^{m}f_{ij}`.
  This method assigns equal weight to each of the possible pairs of focal
  alleles at the two sites, taking the arithmetic mean of statistics over
  focal haplotypes.

Out of all of the available summary functions, only {math}`r^2` uses
`hap_weighted` normalisation, with the remainder using uniform weighting

> (`total_weighted`).

Reviewer: Can we say "see Ragsdale (XXXX) for explanation of this choice" or something?

(sec_stats_two_locus_branch)=

#### Branch

The `branch` mode computes expected two-locus statistics between pairs of
trees, conditioned on the marginal topologies and branch lengths of those
trees. The trees for which we compute statistics are specified by positions,
and for a pair of positions we consider all possible haplotypes that could be

> generated by a single mutation occurring at the two trees.

Reviewer: ```suggestion
generated by a single mutation occurring on each of the two trees.
```

For two trees, one with {math}`n` branches and the other with {math}`m`
branches, there are {math}`nm` possible pairs of branches that may carry the
pair of mutations. For each pair, we compute the two-locus statistic, and then
sum these values weighted by the product of the two branch lengths. Given the
two mutations occur, this accounts for the relative probability that the two
mutations fall on any pair of branches.

> 

Reviewer: What do you think about inserting one of the [precise interpretations](https://github.com/tskit-dev/tskit/pull/3416#issuecomment-4060713700) of what's computed? This is still very vague ("accounts for the relative probability").

Reviewer: Another important thing to mention: which branches? For instance, if there are branches above the root, or not ancestral to any samples, are those included? What about branches ancestral to samples but not to any of the samples in `sample_sets`? I assumed it would be "only branches on which a mutation would produce a polymorphic site", but I'm noticing that I can get `nan`s from branch stats, so maybe not?
```
mts.ld_matrix(sample_sets, stat='r2', indexes=(0,1), mode='branch')
array([[nan, nan, nan],
       [nan, nan, nan],
       [nan, nan, nan]])
```

The time complexity of this method is quadratic in the number of samples,
due to the pairwise comparisons of branches from each pair of trees.
By default, this method computes
a symmetric matrix for all pairs of trees, with rows and columns representing
each tree in the tree sequence. Similar to the site method, we can restrict the
output to a subset of trees, either by specifying a vector of positions or
a pair of vectors for row and column positions separately. To select a specific
tree, the specified positions must land in the tree span (`[start, end)`).

In the following, we compute a matrix of expected {math}`r^2` within and
between the first 4 trees in the tree sequence. The tree breakpoints are
a convenient way to specify those first four trees.

```{code-cell} ipython3
ld = ts.ld_matrix(mode="branch", positions=[ts.breakpoints(as_array=True)[0:4]])
print(ld)
```

Again, we can specify the row and column trees separately.

```{code-cell} ipython3
breakpoints = ts.breakpoints(as_array=True)
ld = ts.ld_matrix(mode="branch", positions=[breakpoints[[0]], breakpoints[0:4]])
print(ld)
```

(sec_stats_two_locus_sample_sets)=

#### Sample Sets

> 

Reviewer: I think we need to first insert a paragraph explaining what happens without `indexes`: i.e., a list of samples gets you a matrix; and a list-of-lists of samples gets you a 3D array. Then say "okay and for a two-way thing you need a list-of-list of indexes, ...".

The two-locus API provides a mechanism by which to subset the samples under
consideration, providing the ability to compute a separate LD matrix for each

> sample set or an LD matrix between sample sets. Output dimensions are handled

Reviewer: This isn't quite right, since separate LD matrices are produced for each sample set by default even for non-two-way statistics.

Reviewer: ```suggestion
consideration, providing the ability to compute a LD matrix between distinct sample sets.
Output dimensions are handled
```

in the same manner as the rest of the stats API (see
{ref}`sec_stats_output_dimensions`). The two-locus API differs from the rest of
the stats API in that we handle one-way and two-way statistics in the same
function call.

> To compute a two-way two-locus statistic, the `index` argument must be

Reviewer: ```suggestion
To compute a two-way two-locus statistic, the `indexes` argument must be
```

provided. The statistics are selected in the same way (with the `stat`
argument), but we provide a restricted set of two-way statistics (see
{ref}`sec_stats_two_locus_summary_functions_two_way`). The dimension-dropping
rules for the result follow the rest of the tskit stats API in that a single list
or tuple will produce a single two-dimensional matrix, while list of these

> will produce a three-dimensional array, with the outer dimension of length

Reviewer: ```suggestion
will produce a three-dimensional array, with the first dimension of length
```

equal to the length of the list.

For example, to compute the {math}`r^2` LD matrix over a subset of samples in
the tree sequence (such as sample nodes 0 through 7), we would specify the
samples as follows:

```
ts = msprime.sim_ancestry(
    20,
    population_size=10000,
    sequence_length=1000,
    recombination_rate=2e-8,
    random_seed=12)
ts = msprime.sim_mutations(ts, rate=2e-8, random_seed=12)

ld = ts.ld_matrix(mode=“site”, sample_sets=range(8))
print(ld)
```

For concreteness, we would expect the following dimensions with the specified
`sample_sets` and `indexes` arguments.

```
# one-way
ts.ld_matrix(sample_sets=None) -> 2 dimensions
ts.ld_matrix(sample_sets=[0, 1, 2, 3]) -> 2 dimensions
ts.ld_matrix(sample_sets=[[0, 1, 2, 3]]) -> 3 dimensions
# two-way
ts.ld_matrix(sample_sets=[[0, 1, 2, 3], [4, 5, 6, 7]], indexes=(0, 1)) -> 2 dimensions
ts.ld_matrix(sample_sets=[[0, 1, 2, 3], [4, 5, 6, 7]], indexes=[(0, 1)]) -> 3 dimensions
```

#### Why are there nan values in the LD matrix?

> For some statistics, it is possible to observe nan entries in the LD matrix,

Reviewer: ```suggestion
For some statistics, it is possible to observe `nan` entries in the LD matrix,
```

> which can be surprising or numerically impact downstream analyses. A nan entry

Reviewer: ```suggestion
which can be surprising or numerically impact downstream analyses. A `nan` entry
```

may occur when computing a ratio statistic (such as {math}`r` or {math}`r^2`)
with a denominator of zero, indicating that one or both sites in the pair are
monomorphic. This can happen for a number of reasons:

- The mutation models allows for reversible mutations, so a back mutation at
  a site has resulted in a single allele despite multiple mutations in the
  history of the sample.
- LD is computed for a subsample of individuals, and some sites are not
  variable among the sample nodes in the subsample.
- A mutation exists above the root of the local tree, so that all samples carry
  the mutation, and one or more sites are not variable.

> 

Reviewer: perhaps address also `nan`s in branch stats here

(sec_stats_two_locus_sample_one_way_stats)=

##### One-way Statistics

One-way statistics are summaries of two loci in a single sample set, using a
triple of haplotype counts {math}`\{w_{Ab}, w_{aB}, w_{AB}\}` and the size of
the sample set {math}`n`, where the capitalized and lowercase letters in our notation
represent alternate alleles.

(sec_stats_two_locus_sample_two_way_stats)=

##### Two-way Statistics

Two-way statistics are summaries of haplotype counts between two sample sets,
which operate on the three haplotype counts (as in one-way stats, above)
computed from each sample set, indexed by `(i, j)`. These statistics take on
a different meaning from their one-way counterparts. For instance {math}`D_i

> D_j` can be thought of as the covariance between two loci when computed for

Reviewer: Isn't this called [`D^2`](https://github.com/tskit-dev/tskit/pull/3416/changes#diff-b01b1c97dca975c3d5b4b674ae04dbc2c1c56305aeaa5c4ca17621d12aaa49b2R981) (also see [here](https://github.com/tskit-dev/tskit/pull/3416/changes#diff-ecbb802115b51a593926a3015f4ff2ec444fa4dc4915af74f4959abc60c84df8R10985)), not `D_i D_j`?

a haplotype in a single sample set.

Only a subset of our summary functions are two-way statistics (see
{ref}`sec_two_locus_summary_functions_two_way`). Note that the unbiased two-way
statistics expect non-overlapping sample sets, and we do not make any
assertions about the sample sets and assume that `i` and `j` represent disjoint
sets of samples (see also the note in {meth}`~TreeSequence.divergence`).

(sec_stats_two_locus_summary_functions)=

#### Summary Functions

(sec_stats_two_locus_summary_functions_one_way)=

##### One-way

The two-locus summary functions all take haplotype counts and sample set size as

> input. Each of our summary functions has the signature

Reviewer: All of these functions are in terms of `p`, not `w`, right? So why introduce `w` at all here?

sequence among all samples. In our computations, row sites are used as the row
("left-hand") loci and column sites are used as the column ("right-hand")
locus, and with a single list of sites specified, we obtain a symmetric square
matrix. 

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[0, 1, 2, 3]])
print(ld)
```

The following demonstrates how we can specify the row and column sites
independently of each other. We're specifying 3 columns and 2 rows, which
computes a subset of the matrix shown above.

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[0, 1], [1, 2, 3]])
print(ld)
```

#### Computational details

Because we allow for two-locus statistics to be computed for multi-allelic
data, we need to be able to combine statistical results from each pair of
alleles into one summary for a pair of sites. This does not affect biallelic
data (and so this section can be skipped on first reading).
We use two implementations for
combining results from multiple alleles: `hap_weighted` and `total_weighted`.
These are statistic-specific and not chosen by the user, with choices motivated
by [Zhao (2007)](https://doi.org/10.1017/S0016672307008634) with the goal

> of return statistics as expected by the user.

Reviewer: ```suggestion
by [Zhao (2007)](https://doi.org/10.1017/S0016672307008634).
```
At least this should say "returning" instead of "return", but the whole phrase is kinda mysterious, so I vote to delete.

depends on whether the statistic is polarised.) Write {math}`f_{i,j}` as the
statistic computed for focal alleles {math}`A_i` and {math}`B_j`. Then the
weighting schemes are defined as:

- `hap_weighted`: {math}`\sum_{i=1}^{n}\sum_{j=1}^{m}p(A_{i}B_{j})f_{ij}`,
  where {math}`p(A_{i}B_{j})` is the frequency of haplotype {math}`A_{i}B_{j}`.
  This method was first introduced in [Karlin
  (1981)](https://doi.org/10.1111/j.1469-1809.1981.tb00308.x) and reviewed in
  [Zhao (2007)](https://doi.org/10.1017/S0016672307008634).

- `total_weighted`: {math}`\frac{1}{n m}\sum_{i=1}^{n}\sum_{j=1}^{m}f_{ij}`.
  This method assigns equal weight to each of the possible pairs of focal
  alleles at the two sites, taking the arithmetic mean of statistics over
  focal haplotypes.

Out of all of the available summary functions, only {math}`r^2` uses
`hap_weighted` normalisation, with the remainder using uniform weighting
(`total_weighted`).

Within this framework, statistics may be either
polarised or unpolarised. For statistics that are polarised, we compute
statistic values for pairs of derived alleles. Unpolarised statistics compute

> statistics over all pairs of alleles, derived and ancestral. In either case,

Reviewer: ```suggestion
statistic values for pairs of derived alleles. (For this purpose, the "derived" alleles
at a site are all alleles except that stored as the ``ancestral_state`` for the site.)
Unpolarised statistics compute
statistics over all pairs of alleles, derived and ancestral. In either case,
```

the result is averaged over these values, using a weighting
scheme described below. The option for polarisation is not exposed to the user.

> Instead, we implement statistics that are polarised where appropriate.

Reviewer: ```suggestion
scheme described below. The option for polarisation is not exposed to the user,
and list which statistics are polarised below.
```

(sec_stats_two_locus_branch)=

#### Branch mode

The `branch` mode computes expected two-locus statistics between pairs of
trees, conditioned on the marginal topologies and branch lengths of those
trees. The trees for which we compute statistics are specified by positions,
and for a pair of positions we consider all possible haplotypes that could be
generated by a single mutation occurring on each of the two trees.

For two trees, one with {math}`n` branches and the other with {math}`m`
branches, there are {math}`nm` possible pairs of branches that may carry the
pair of mutations. For each pair, we compute the two-locus statistic, and then
sum these values weighted by the product of the two branch lengths. Given the
two mutations occur, this accounts for the relative probability that the two
mutations fall on any pair of branches.

In other words, imagine we place two mutations uniformly, one on each tree, and
then compute the statistic; the branch mode computes the expected value of the
statistic over this process, multiplied by the product of the total branch
lengths of each tree. In some cases, this process will compute statistics over

> branches that do not result in a polymorphism in the sample, in which case

Reviewer: ```suggestion
lengths of each tree. This weighting accounts for mutational opportunity, so that
the sum of the branch-mode statistic over all positions in a genomic region,
multiplied by a mutation rate,  is equal to the expected sum of the two-locus site
statistic over all mutations falling in that region under an infinite-sites model.

If some branches are ancestral to none or all of the samples, then a mutation on
these branches does not result in a polymorphism in the sample, and so the
```

ratio statistics ({math}`r`, {math}`r^2`)  will return `nan` values. See

> further discussion of `nan` values below.

Reviewer: Actually, I think this bit on `nan` should just be moved to the section below.

Author: This information was also repeated in that section below, so I've moved/condensed it.

indexes of the `sample_sets`, which must be a list of lists of sample nodes.
This results in an LD matrix computed for each list of indexes. The statistics
are selected in the same way (with the `stat` argument), and these are limited
to a handful of statistics (see
{ref}`sec_stats_two_locus_summary_functions_two_way`). The dimension-dropping
rules for the result follow the rest of the tskit stats API in that a single
list or tuple will produce a single two-dimensional matrix, while list of these
will produce a three-dimensional array, with the first dimension of length
equal to the length of the list.

For example, to compute the {math}`r^2` LD matrix over a subset of samples in
the tree sequence (such as sample nodes 0 through 7), we would specify the
samples as follows:

```
ts = msprime.sim_ancestry(
    20,
    population_size=10000,
    sequence_length=1000,
    recombination_rate=2e-8,
    random_seed=12)
ts = msprime.sim_mutations(ts, rate=2e-8, random_seed=12)

ld = ts.ld_matrix(mode=“site”, sample_sets=range(8))
print(ld)
```

For concreteness, we would expect the following dimensions with the specified

> `sample_sets` and `indexes` arguments.

Reviewer: ```suggestion
We would get the following dimensions with the specified
`sample_sets` and `indexes` arguments.
```

```
# one-way
ts.ld_matrix(sample_sets=None) -> 2 dimensions
ts.ld_matrix(sample_sets=[0, 1, 2, 3]) -> 2 dimensions
ts.ld_matrix(sample_sets=[[0, 1, 2, 3]]) -> 3 dimensions
# two-way
ts.ld_matrix(sample_sets=[[0, 1, 2, 3], [4, 5, 6, 7]], indexes=(0, 1)) -> 2 dimensions

> ts.ld_matrix(sample_sets=[[0, 1, 2, 3], [4, 5, 6, 7]], indexes=[(0, 1)]) -> 3 dimensions

Reviewer: ```suggestion
ts.ld_matrix(sample_sets=None) # -> 2 dimensions
ts.ld_matrix(sample_sets=[0, 1, 2, 3]) # -> 2 dimensions
ts.ld_matrix(sample_sets=[[0, 1, 2, 3]]) # -> 3 dimensions
# two-way
ts.ld_matrix(sample_sets=[[0, 1, 2, 3], [4, 5, 6, 7]], indexes=(0, 1)) # -> 2 dimensions
ts.ld_matrix(sample_sets=[[0, 1, 2, 3], [4, 5, 6, 7]], indexes=[(0, 1)]) # -> 3 dimensions
```

```

#### Why are there `nan` values in the LD matrix?

For some statistics, it is possible to observe `nan` entries in the LD matrix,
which can be surprising or numerically impact downstream analyses. A `nan` entry
may occur when computing a ratio statistic (such as {math}`r` or {math}`r^2`)
with a denominator of zero, indicating that one or both sites in the pair are

> monomorphic. This can happen for a number of reasons:

Reviewer: ```suggestion
occurs if the denominator of a ratio statistic (such as {math}`r` or {math}`r^2`)
is zero, indicating that one or both of the alleles in the pair is fixed or absent in the
sample set under consideration. This can happen for a number of reasons:
```

- The mutation models allows for reversible mutations, so a back mutation at
  a site has resulted in a single allele despite multiple mutations in the
  history of the sample.
- LD is computed for a subsample of individuals, and some sites are not
  variable among the sample nodes in the subsample.
- A mutation exists above the root of the local tree, so that all samples carry
  the mutation, and one or more sites are not variable.

The `branch` mode will also return `nan` values if there are branches in either
tree on which a mutation would not result in a polymorphism within a sample

> set.

Reviewer: I think we would like some more discussion of this, but we can push tihs to a different PR: see #3441.

(sec_stats_two_locus_sample_one_way_stats)=

##### One-way Statistics

One-way statistics are summaries of two loci in a single sample set, using
a triple of haplotype counts {math}`\{n_{Ab}, n_{aB}, n_{AB}\}` and the size of
the sample set {math}`n`, where the capitalized and lowercase letters in our
notation represent alternate alleles.

(sec_stats_two_locus_sample_two_way_stats)=

##### Two-way Statistics

Two-way statistics are summaries of haplotype counts between two sample sets,
which operate on the three haplotype counts (as in one-way stats, above)
computed from each sample set, indexed by `(i, j)`. These statistics take on

> a different meaning from their one-way counterparts. For instance {math}`D2`

Reviewer: ```suggestion
a different meaning from their one-way counterparts. For instance `stat="D2"`
```
(right?)

over a pair of sample sets computes {math}`D_i D_j`, which is the product of
the covariance measure of LD within each sample set and is related to the
covariance of {math}`D` between sample sets.

Only a subset of our summary functions are two-way statistics (see
{ref}`sec_two_locus_summary_functions_two_way`). Note that the unbiased two-way
statistics expect non-overlapping sample sets, and we do not make any
assertions about the sample sets and assume that `i` and `j` represent disjoint
sets of samples (see also the note in {meth}`~TreeSequence.divergence`).

(sec_stats_two_locus_summary_functions)=

#### Summary Functions

(sec_stats_two_locus_summary_functions_one_way)=

##### One-way

The two-locus summary functions all take haplotype counts and sample set size
as input. Each of our summary functions has the signature
{math}`f(n_{Ab}, n_{aB}, n_{AB}, n)`, converting to haplotype frequencies

> {math}`\{p_{Ab}, p_{aB}, p_{AB}\}` where appropriate by dividing by {math}`n`.

Reviewer: Which alleles are `A` and which are `a`? My guess is that for polarised statistics, `A` is non-ancestral, while for unpolarised it doesn't matter because we sum over all *unordered* pairs (but we should say so).

Reviewer: hm, looking at how the stats mostly use `pa` not `pA` I wonder if `a` is derived?

Reviewer: > we sum over all unordered pairs 

I mean *average* over all unordered pairs, using the weighting scheme.

Author: I found and fixed additional issues with this section of the docs, especially the `pa` vs `pA` mixup. I also added a note about the labels `A,B` and `a,b`.

> {math}`\{p_{Ab}, p_{aB}, p_{AB}\}` where appropriate by dividing by {math}`n`.

Reviewer: ```suggestion
{math}`\{p_{Ab}, p_{aB}, p_{AB}\}` by dividing by {math}`n`.
```

`D`
: {math}`f(n_{Ab}, n_{aB}, n_{AB}, n) = p_{ab} - p_{a}p_{b}`

  This statistic is polarised, as the unpolarised result, which averages over
  allele labelings, is zero. Uses the `total` weighting method.

`D_prime`
: {math}`f(n_{Ab}, n_{aB}, n_{AB}, n) = \frac{D}{D_{\max}}`,

  where {math}
            \min\{p_A (1-p_B), p_B (1-p_B)\} & \textrm{if }D>=0 \\
            \min\{p_A p_B, (1-p_B) (1-p_B)\} & \textrm{otherwise}
        \end{cases}```

>   and {math}`D` is defined above. Polarised, `total` weighted.

Reviewer: what is Dmax?

Reviewer: Oh, I see - this bit got messed up somehow, I think it's missing the `Dmax = \begin{cases} ...`

similarly to most other functions with respect to `sample_sets` and `indexes`.

Two-locus statistics can be computed using two {ref}`modes <sec_stats_mode>`,
either `site` or `branch`, and these should be interpreted in the same way as
these modes in the single-site statistics. That is, the `site` mode computes LD
over observed alleles at pairs of sites, while the `branch` model computes
expected LD conditioned on pairs of trees.

(sec_stats_two_locus_site)=

#### Site mode

The `"site"` mode computes two-locus statistics summarized over alleles between
all pairs of specified sites. The default behavior, leaving `sites`
unspecified, will compute a matrix for all pairs of sites, with one row and
column for each site in the tree sequence (i.e., an {math}`n \times n` matrix
where {math}`n` is the number of sites in the tree sequence). We can also
restrict the output to a subset of sites, either by specifying a single vector
of site indexes for both rows and columns or a pair of vectors for the row
sites and column sites separately.

The following computes a matrix of the {math}`r^2` measure of linkage
disequilibrium (LD) computed pairwise between the first 4 sites in the tree

> sequence among all samples. The `sites` must be given as a list of lists, and
> with a single list of sites specified, we obtain a symmetric square matrix. 

Reviewer: ```suggestion
sequence among all samples. The `sites` must be given as either a list
or a list of lists of sample IDs.
With a single list of sites specified, we obtain a symmetric square matrix. 
```

Author: With a unnested list for `sites`, such as `ts.ld_matrix(sites=[0, 1, 2, 3])`, I get `ValueError: Sites must be a list of lists, tuples, or ndarrays`

Reviewer: Whoops, my mistake! Sorry.

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[0, 1, 2, 3]])
print(ld)
```

The following demonstrates how we can specify the row and column sites

> independently of each other. We're specifying 2 rows and 3 columns, which

Reviewer: ```suggestion
If a list of two lists of sample IDs is provided, these specify the row and column sites.
For instance, here we're specifying 2 rows and 3 columns, which
```

Author: I think you mean if a "list of two lists of site indexes is provided," but I see how this writing can be improved.

Reviewer: You're right; as it's written is actually better. Never mind.

computes a subset of the matrix shown above.

```{code-cell} ipython3
ld = ts.ld_matrix(sites=[[1, 2], [1, 2, 3]])
print(ld)
```

#### Computational details

Because we allow for two-locus statistics to be computed for multi-allelic
data, we need to be able to combine statistical results from each pair of
alleles into one summary for a pair of sites. This does not affect biallelic
data (and so this section can be skipped on first reading).
We use two implementations for
combining results from multiple alleles: `hap_weighted` and `total_weighted`.
These are statistic-specific and not chosen by the user, with choices motivated
by [Zhao (2007)](https://doi.org/10.1017/S0016672307008634).

Briefly, consider a pair of sites with {math}`n` alleles at the first locus and
{math}`m` alleles at the second. (Whether this includes the ancestral allele

> depends on whether the statistic is polarised.) Write {math}`f_{i,j}` as the

Reviewer: ```suggestion
depends on whether the statistic is polarised.) Write {math}`f_{ij}` as the
```

each tree in the tree sequence. Similar to the site method, we can restrict the
output to a subset of trees, either by specifying a vector of positions or
a pair of vectors for row and column positions separately. To select a specific
tree, the specified positions must land in the tree span (`[start, end)`).

In the following, we compute a matrix of expected {math}`r^2` within and
between the first 4 trees in the tree sequence. The tree breakpoints are
a convenient way to specify those first four trees.

```{code-cell} ipython3
ld = ts.ld_matrix(
    mode="branch",
    positions=[ts.breakpoints(as_array=True)[0:4]]
)
print(ld)
```

We note that these values are quite large: as described above, the statistic is
scaled by the product of the total branch lengths of each pair of trees. To
compute the expected {math}`r^2` value for a pair of mutations that each land
uniformly on the pair of trees, we can divide by the product of the total
branch lengths:

```{code-cell} ipython3
total_branch_lengths = [tree.total_branch_length for tree in ts.trees()]
prod_branch_lengths = np.outer(total_branch_lengths, total_branch_lengths)
print(ld / prod_branch_lengths[0:4, 0:4])
```

> 

Reviewer: I think this is right, and I think could help make things concrete?
```suggestion

To compute the average {math}`r^2` for a uniformly chosen pair of mutations, we also
weight by tree span:
```{code-cell}
tree_spans = np.array([t.span for t in ts.trees()])
total_opportunity = np.sum(tree_spans * total_branch_lengths)
all_ld = ts.ld_matrix(mode="branch")
mean_ld = np.sum( all_ld * np.outer(tree_spans, tree_spans)) / total_opportunity**2
print("mean infinite-sites LD:", mean_ld)
```

```

{math}`r^2`) is zero, indicating that one or both of the alleles in the pair is
fixed or absent in the given sample set(s). This can happen for
a number of reasons:

- Some mutation models allow for reversible mutations, so a back mutation at
  a site can result in a single allele despite multiple mutations in the
  history of the sample.
- LD is computed for a subsample of individuals, and some sites are not
  variable among the sample nodes in the subsample.
- A mutation exists above the root of the local tree, so that all samples carry
  the mutation, and one or more sites are not variable.

The `branch` mode will also return `nan` values for ratio statistics if there
are branches in either tree on which a mutation would not result in
a polymorphism within a sample set.

(sec_stats_two_locus_sample_one_way_stats)=

#### One-way Statistics

One-way statistics are summaries of two loci in a single sample set, using
a triple of haplotype counts {math}`\{n_{AB}, n_{Ab}, n_{aB}\}` and the size of
the sample set {math}`n`, where the capitalized and lowercase letters in our
notation represent alternate alleles.

> 

Reviewer: ```suggestion
::::{warning}
This means there are two common situations in which many or all LD values will be nan.
These are:

1. A branch-mode ratio statistic computed on less than the full set of samples
    will always be nan, since part of the trees are ancestral to none of the samples.
2. A site-mode ratio statistic will be nan at any sites at which there are alleles found
    in the entire set of samples that are not seen in the provided sample set.

This behavior **may change in the future**,
because possibly more natural behavior not currently implemented
would be to ignore the branches/alleles not ancestral
to any of the provided samples.
```

  and {math}`v_j` is the covariance of the trait with the j-th covariate.

(sec_stats_multi_site)=

## Multi site statistics

(sec_stats_two_locus)=

### Two-locus statistics

The {meth}`~TreeSequence.ld_matrix` method provides an interface to
a collection of two-locus statistics with predefined summary functions (see
{ref}`sec_stats_two_locus_summary_functions`) and `site` and `branch`
{ref}`modes <sec_stats_mode>`. The LD matrix method differs from other
statistics methods in that it provides a unified API with an argument to
specify different two-locus summaries of the data. It otherwise behaves
similarly to most other functions with respect to `sample_sets` and `indexes`.

> Two-locus statistics can be computed using two modes, either `sites` or

Reviewer 2: ```suggestion
Two-locus statistics can be computed using two modes, either `site` or
```
